diff -r 72720e9dbb41 platform/60D.111/Makefile.platform.default
--- a/platform/60D.111/Makefile.platform.default	Fri Mar 27 00:37:38 2015 +0200
+++ b/platform/60D.111/Makefile.platform.default	Sun Mar 29 23:40:21 2015 +0300
@@ -12,7 +12,7 @@
 RESTARTSTART  = 0x0005f000
 
 ifeq ($(ML_SRC_PROFILE),generic)
-RESTARTSTART  = 0x1FE00000
+#~ RESTARTSTART  = 0x1FE00000
 endif
 
 # must have 8.3 characters for 64GB EXFAT cards
diff -r 72720e9dbb41 platform/60D.111/consts.h
--- a/platform/60D.111/consts.h	Fri Mar 27 00:37:38 2015 +0200
+++ b/platform/60D.111/consts.h	Sun Mar 29 23:40:21 2015 +0300
@@ -15,9 +15,9 @@
 #define HIJACK_INSTR_MY_ITASK 0xFF0110DC
 #define HIJACK_TASK_ADDR 0x1a2c
 
-#define HIJACK_CACHE_HACK
-#define HIJACK_CACHE_HACK_INITTASK_ADDR  0xFF0110DC
-#define RSCMGR_MEMORY_PATCH_END 0xF8013698
+//~ #define HIJACK_CACHE_HACK
+//~ #define HIJACK_CACHE_HACK_INITTASK_ADDR  0xFF0110DC
+//~ #define RSCMGR_MEMORY_PATCH_END 0xF8013698
 
 #define ARMLIB_OVERFLOWING_BUFFER 0x36468 // in AJ_armlib_setup_related3
 
diff -r 72720e9dbb41 platform/60D.111/features.h
--- a/platform/60D.111/features.h	Fri Mar 27 00:37:38 2015 +0200
+++ b/platform/60D.111/features.h	Sun Mar 29 23:40:21 2015 +0300
@@ -1,12 +1,1 @@
-#include "all_features.h"
-
-#define FEATURE_SWAP_MENU_ERASE
-#define FEATURE_EYEFI_TRICKS
-#define FEATURE_REC_NOTIFY_BEEP
-
-#undef FEATURE_UPSIDE_DOWN // do we really need it with the swivel screen?
-#undef FEATURE_LV_ZOOM_AUTO_EXPOSURE // not working well anyway
-
-#undef FLEXINFO_DEVELOPER_MENU
-
-#undef FEATURE_WIND_FILTER // problems with external mic
+#define FEATURE_SCREENSHOT
diff -r 72720e9dbb41 platform/60D.111/internals.h
--- a/platform/60D.111/internals.h	Fri Mar 27 00:37:38 2015 +0200
+++ b/platform/60D.111/internals.h	Sun Mar 29 23:40:21 2015 +0300
@@ -108,14 +108,14 @@
 #define CONFIG_EVF_STATE_SYNC
 
 /** We have access to raw data in LiveView and Photo mode */
-#define CONFIG_RAW_LIVEVIEW
-#define CONFIG_RAW_PHOTO
+//~ #define CONFIG_RAW_LIVEVIEW
+//~ #define CONFIG_RAW_PHOTO
 
 /** We know how to use edmac_memcpy */
-#define CONFIG_EDMAC_MEMCPY
+//~ #define CONFIG_EDMAC_MEMCPY
 
 /** this method bypasses Canon's lv_save_raw and slurps the raw data directly from connection #0 */
-#define CONFIG_EDMAC_RAW_SLURP
+//~ #define CONFIG_EDMAC_RAW_SLURP
 
 /** Use a patched LiveViewApp dialog hander to hide Canon bottom bar */
 #define CONFIG_LVAPP_HACK_RELOC
diff -r 72720e9dbb41 src/Makefile.src
--- a/src/Makefile.src	Fri Mar 27 00:37:38 2015 +0200
+++ b/src/Makefile.src	Sun Mar 29 23:40:21 2015 +0300
@@ -171,6 +171,7 @@
 	edmac-memcpy.o \
 	patch.o \
 	patch-simple.o \
+	asm.o \
 
 
 ML_OBJS-y += $(ML_SRC_OBJS)
diff -r 72720e9dbb41 src/config-defines.h
--- a/src/config-defines.h	Fri Mar 27 00:37:38 2015 +0200
+++ b/src/config-defines.h	Sun Mar 29 23:40:21 2015 +0300
@@ -22,20 +22,20 @@
  */
 
     /** If something goes wrong (ERR70), we can save a crash log **/
-    #define CONFIG_CRASH_LOG
+    //~ #define CONFIG_CRASH_LOG
 
     /** It's a good idea to back up ROM contents on the card - just in case **/
-    #define CONFIG_AUTOBACKUP_ROM
+    //~ #define CONFIG_AUTOBACKUP_ROM
 
     /** It's a good idea to run some automated tests **/
-    #define CONFIG_STRESS_TEST
-    #define CONFIG_BENCHMARKS
+    //~ #define CONFIG_STRESS_TEST
+    //~ #define CONFIG_BENCHMARKS
 
     /** You may want to disable this for troubleshooting **/
     #define CONFIG_CONFIG_FILE
     
     /** Show detailed info about tasks and CPU usage */
-    #define CONFIG_TSKMON
+    //~ #define CONFIG_TSKMON
 
 /**
  * Some debug stuff - you should enable it Makefile.user to avoid pushing unwanted changes to the repo
